File: scala/155.scala
import scala.collection.mutable.Stack
class MinStack() {

    /** initialize your data structure here. */
  var s = Stack[Int]()
  var minStack = Stack[Int]()

    def push(`val`: Int) {
      s.push(`val`)
      if (minStack.isEmpty || `val` <= minStack.top) minStack.push(`val`)
    }

    def pop() {
      val v = s.pop()
      if (!minStack.isEmpty && minStack.top == v) minStack.pop()
    }

    def top(): Int = {
        s.top
    }

    def getMin(): Int = {
        minStack.top
    }

}

File: leetcodeJava/155.java
package leetcodeJava;

class MinStack {
    LinkedList<Integer> stack = new LinkedList<Integer>();
    LinkedList<Integer> minStack = new LinkedList<Integer>();

    public MinStack() {
        
    }
    
    public void push(int val) {
        if (stack.isEmpty()) {
            stack.add(new Integer(val));
            minStack.add(new Integer(val));
        } else {
            if (val <= minStack.getLast()) {
                minStack.add(new Integer(val));
            }
            stack.add(new Integer(val));
        }
    }
    
    public void pop() {
        if (stack.getLast().equals(minStack.getLast())) {
            stack.removeLast();
            minStack.removeLast();
        } else {
            stack.removeLast();
        }
    }
    
    public int top() {
        return stack.getLast();
    }
    
    public int getMin() {
        return minStack.getLast();
    }
}
